test/eigsolve.jl

```diff
@@ -359,3 +359,277 @@ end
     @test_logs realeigsolve(A, v, 1, :LM, Arnoldi(; tol=1e-12, verbosity=1))
     @test_logs (:info,) realeigsolve(A, v, 1, :LM, Arnoldi(; tol=1e-12, verbosity=2))
 end
+
+@testset "BlockLanczos - eigsolve for large sparse matrix and map input" begin
+    function toric_code_strings(m::Int, n::Int)
+        li = LinearIndices((m, n))
+        bottom(i, j) = li[mod1(i, m), mod1(j, n)] + m * n
+        right(i, j) = li[mod1(i, m), mod1(j, n)]
+        xstrings = Vector{Int}[]
+        zstrings = Vector{Int}[]
+        for i in 1:m, j in 1:n
+            # face center
+            push!(xstrings, [bottom(i, j - 1), right(i, j), bottom(i, j), right(i - 1, j)])
+            # cross
+            push!(zstrings, [right(i, j), bottom(i, j), right(i, j + 1), bottom(i + 1, j)])
+        end
+        return xstrings, zstrings
+    end
+
+    function pauli_kron(n::Int, ops::Pair{Int,Char}...)
+        mat = sparse(1.0I, 2^n, 2^n)
+        for (pos, op) in ops
+            if op == 'X'
+                σ = sparse([0 1; 1 0])
+            elseif op == 'Y'
+                σ = sparse([0 -im; im 0])
+            elseif op == 'Z'
+                σ = sparse([1 0; 0 -1])
+            elseif op == 'I'
+                σ = sparse(1.0I, 2, 2)
+            else
+                error("Unknown Pauli operator $op")
+            end
+
+            left = sparse(1.0I, 2^(pos - 1), 2^(pos - 1))
+            right = sparse(1.0I, 2^(n - pos), 2^(n - pos))
+            mat = kron(left, kron(σ, right)) * mat
+        end
+        return mat
+    end
+
+    # define the function to construct the Hamiltonian matrix
+    function toric_code_hamiltonian_matrix(m::Int, n::Int)
+        xstrings, zstrings = toric_code_strings(m, n)
+        N = 2 * m * n  # total number of qubits
+
+        # initialize the Hamiltonian matrix as a zero matrix
+        H = spzeros(2^N, 2^N)
+
+        # add the X-type operator terms
+        for xs in xstrings[1:(end - 1)]
+            ops = [i => 'X' for i in xs]
+            H += pauli_kron(N, ops...)
```

Comment:
Here too I would avoid the intermediate construction of vectors and simply write
```suggestion
            H += pauli_kron(N, (i => 'X' for i in xs)...)
```